### PR TITLE
Basic entity pathfinding

### PR DESCRIPTION
**Pull request by boy0001**
Hostile mobs will traverse around obstacles to attack the player (only hostile mob is creeper right now)

Example usage:

``` Java
MobAI a = new NeutralMobAI(this, 16, 16, 0);
MobAI b = new FollowOwnerAI(this, 128);
MobAI c = new FollowItemAI(this, Item.BONE, 32, 16);
MobAI d = new CreatureWanderAI(this);
entity.setAI(new MobAIUnion(a, b, c, d));
```

### CODE REVIEW

**Comment by yescallop** [comment]
Bugs found:
 - Wolf causes client crashing
 - Entity sometimes stuck in a corner

Details:
 - Entity cannot float on water

**Comment by boy0001** [comment]
@yescallop 
Jump height was too low - fixed
Seems the client doesn't recognize the nbt related to tamed entities, so I commented that out

**Comment by yescallop** [comment]
Jump height too high that jumping onto a fence is possible

**Comment by yescallop** [comment]
Knocking back an entity looks strange

**Comment by yescallop** [comment]
Falling is no damage

**Comment by boy0001** [comment]
> Entity cannot float on water
Fixed

> Jump height too high that jumping onto a fence is possible
The pathfinder should determine where it jumps, so jump height shouldn't matter.

> Knocking back an entity looks strange
It's always looked strange

> Falling is no damage
Yes there is?

**Comment by yescallop** [comment]
https://github.com/Nukkit/Nukkit/pull/1149/commits/ab56577f2ff51adb3ef13dd17489319e18c28df8
Seems this causes a crash

**Comment by boy0001** [comment]
> Seems this commit causes a client crash

That commit just changes the motionX/Y/Z values for the Entity, it wouldn't make sense for it to cause a crash, and it's not crashing at all for me?

If you have any wolves from earlier you need to delete them as Nukkit's EntityTameable class caused the crash there and I had to comment it out.

**Comment by yescallop** [comment]
But you've fixed it, confirmed the crash isn't caused by https://github.com/Nukkit/Nukkit/pull/1149/commits/ab56577f2ff51adb3ef13dd17489319e18c28df8

**Comment by kniffo80** [comment]
The tamed stuff was copied from original PC server version. Just checked that and tried a lot of stuff with taming. It seems, that the data structure for MCPE is another one than on PC version. I took a look on a lot of other implementations - but none found out how taming is implemented.

A decompiled android client would be a good idea (but i think only the library which is - i assume at this point - written in C++). Then we can take a look how it's implemented.

File: src/main/java/cn/nukkit/entity/ai/MobAIUnion.java [hunk 1]
```diff
@@ -0,0 +1,55 @@
+package cn.nukkit.entity.ai;
+
+import cn.nukkit.entity.Entity;
+import cn.nukkit.math.Vector3;
+
+/**
+ * You can combine different AI's together
+ *  - The fist MobAI to return a valid target is used
+ */
+public class MobAIUnion implements MobAI {
```
Comment (liach) [inline src/main/java/cn/nukkit/entity/ai/MobAIUnion.java]: Well I assume that you can create some class like `MobAISet` for the AI for a mob.

**Comment by Creeperface01** [comment]
@boy0001 awesome! but why did you add getFloorX() to block class? its already in Vector3 class :D

**Comment by IrishPacks** [comment]
I seem to be getting errors.

22:48:05 [ALERT] java.lang.NullPointerException
        at java.io.Reader.<init>(Unknown Source)
        at java.io.InputStreamReader.<init>(Unknown Source)
        at cn.nukkit.utils.Utils.readFile(Utils.java:72)
        at cn.nukkit.lang.BaseLang.loadLang(BaseLang.java:88)
        at cn.nukkit.lang.BaseLang.<init>(BaseLang.java:37)
        at cn.nukkit.lang.BaseLang.<init>(BaseLang.java:29)
        at cn.nukkit.lang.BaseLang.<init>(BaseLang.java:25)
        at cn.nukkit.Server.<init>(Server.java:278)
        at cn.nukkit.Nukkit.main(Nukkit.java:65)

**Comment by IrishPacks** [comment]
One problem I find with this is that the mobs move too fast and don't look the direction they're going and oh boy do they move far when you hit them. but this is just the basics so good job :)

**Comment by boy0001** [comment]
@Creeperface01 
The getFloorX/Z was broken for entities (returned wrong value for negative coords). I fixed it, but figured it would be faster to just cast to int for blocks since they always have an integer value.

@IrishPacks 
Error isn't related to this at all. In the nukkit.yml reset the `lang` to `eng` or something, or just delete the whole file and it should fix it.

And yeah, I haven't touched the default speed for entities (not really sure what the right values would have been).

**Comment by IrishPacks** [comment]
Yeah i fixed the lang issues, I'll mess around with the speed values and get back to you :)

**Comment by Creeperface01** [comment]
Use values from pureentities

**Comment by Creeperface01** [comment]
so when this will be ready to merge? :D

**Comment by File14** [comment]
When will this be ready? @boy0001

**Comment by Creeperface01** [comment]
@boy0001 can you update it? :D

**Comment by Creeperface01** [comment]
@boy0001 or if you could create patch it would be good

**Comment by Creeperface01** [comment]
im testing it and looks good, only mobs speed is too high :D and move duration too short

**Comment by Creeperface01** [comment]
so its only about values now i think

**Comment by Creeperface01** [comment]
maybe entity.getSpeed() / 20 because speed is in blocks per second

**Comment by File14** [comment]
It would be cool if we could add this! Finally moveable mobs :D

**Comment by PeratX** [comment]
Outdated.

**Pull request closed without merging**